backend/coins/eth/account.go

```diff
@@ -285,19 +287,41 @@ type TxProposal struct {
 
 func (account *Account) newTx(
 	recipientAddress string,
-	amount coin.SendAmount) (*TxProposal, error) {
+	amount coin.SendAmount,
+	data []byte,
+) (*TxProposal, error) {
 	if !common.IsHexAddress(recipientAddress) {
 		return nil, errp.WithStack(coin.ErrInvalidAddress)
 	}
-	const gasLimit = 21000 // simple transaction gas cost
 
 	suggestedGasPrice, err := account.coin.client.SuggestGasPrice(context.TODO())
 	if err != nil {
 		return nil, err
 	}
-	fee := new(big.Int).Mul(big.NewInt(gasLimit), suggestedGasPrice)
 
-	var value *big.Int
+	parsedAmount, err := amount.Amount(big.NewInt(params.Ether))
+	if err != nil {
+		return nil, err
+	}
+	value := parsedAmount.BigInt()
+
+	address := common.HexToAddress(recipientAddress)
+	message := ethereum.CallMsg{
+		From:     account.address.Address,
+		To:       &address,
+		Gas:      0,
+		GasPrice: suggestedGasPrice,
+		Value:    value,
+		Data:     data,
+	}
+
+	gasLimit, err := account.coin.client.EstimateGas(context.TODO(), message)
+	if err != nil {
+		return nil, err
```

Comment:
this error happens easily if you add a contract address which fails to execute with the input. should return an error code for the frontend to show a good message.